File: src/collector/news_collector.py
"""
뉴스 수집 모듈

네이버 경제/정치 뉴스 크롤링 및 글로벌 RSS 피드 수집
상태를 가지지 않으므로 모듈 함수로 구성
"""

import asyncio
from datetime import datetime

import feedparser
import httpx
from bs4 import BeautifulSoup
from loguru import logger

from src.common.types import NewsArticle

# 뉴스 카테고리 분류용 키워드 매핑
_CATEGORY_KEYWORDS: dict[str, list[str]] = {
    "policy": ["금리", "기준금리", "한은", "연준", "fed", "통화정책", "인플레"],
    "politics": ["국회", "대통령", "정부", "총리", "외교", "북한", "선거"],
    "industry": ["반도체", "ai", "배터리", "자동차", "it", "테크", "기술"],
    "market": ["코스피", "코스닥", "나스닥", "s&p", "dow", "증시", "주가"],
    "global": ["미국", "중국", "일본", "유럽", "글로벌", "세계"],
    "trend": ["트렌드", "mz", "소비", "유행"],
}

# 글로벌 RSS 피드 소스
_RSS_FEEDS = [
    {
        "url": "https://feeds.bbci.co.uk/news/business/rss.xml",
        "source": "BBC Business",
        "category": "global",
    },
    {
        "url": "https://rss.nytimes.com/services/xml/rss/nyt/Economy.xml",
        "source": "NYT Economy",
        "category": "global",
    },
    {
        "url": "https://search.cnbc.com/rs/search/combinedcms/view.xml?partnerId=wrss01&id=10001147",
        "source": "CNBC",
        "category": "global",
    },
]

_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36"
}


def _categorize_news(text: str) -> str:
    """뉴스 제목+내용으로 카테고리를 자동 분류

    Args:
        text: 분류할 텍스트 (제목 + 요약)

    Returns:
        뉴스 카테고리 문자열
    """
    lower = text.lower()
    for category, keywords in _CATEGORY_KEYWORDS.items():
        if any(kw in lower for kw in keywords):
            return category
    return "economy"


async def _fetch_naver_section_news(
    section_id: str, section_name: str, default_category: str
) -> list[NewsArticle]:
    """네이버 뉴스 특정 섹션의 헤드라인을 크롤링

    Args:
        section_id: 네이버 뉴스 섹션 ID (100=정치, 101=경제)
        section_name: 섹션 표시 이름
        default_category: 분류 실패 시 기본 카테고리

    Returns:
        수집된 뉴스 기사 목록
    """
    url = f"https://news.naver.com/section/{section_id}"
    articles: list[NewsArticle] = []

    try:
        async with httpx.AsyncClient(timeout=10) as client:
            response = await client.get(url, headers=_HEADERS)
            response.raise_for_status()

        soup = BeautifulSoup(response.text, "html.parser")

        for item in soup.select("div.sa_text"):
            title_el = item.select_one("a.sa_text_title")
            if not title_el:
                continue

            strong = title_el.select_one("strong")
            title = strong.get_text(strip=True) if strong else ""
            href = title_el.get("href", "")
            lede = item.select_one("div.sa_text_lede")
            summary = lede.get_text(strip=True)[:200] if lede else ""

            if title:
                category = (
                    default_category
                    if default_category == "politics"
                    else _categorize_news(f"{title} {summary}")
                )
                articles.append(
                    NewsArticle(
                        title=title,
                        summary=summary,
                        url=str(href),
                        source=f"네이버 {section_name}",
                        published_at=datetime.now().isoformat(),
                        category=category,
                    )
                )

        logger.info(f"네이버 {section_name} 뉴스 {len(articles)}건 수집")
    except Exception as e:
        logger.error(f"네이버 {section_name} 뉴스 수집 실패: {e}")

    return articles[:15]


async def _fetch_rss_news() -> list[NewsArticle]:
    """RSS 피드에서 글로벌 경제 뉴스를 수집

    Returns:
        수집된 글로벌 뉴스 기사 목록
    """
    articles: list[NewsArticle] = []

    for feed_config in _RSS_FEEDS:
        try:
            parsed = await asyncio.to_thread(feedparser.parse, feed_config["url"])
            entries = parsed.entries[:5]

            for entry in entries:
                articles.append(
                    NewsArticle(
                        title=entry.get("title", ""),
                        summary=entry.get("summary", "")[:200],
                        url=entry.get("link", ""),
                        source=feed_config["source"],
                        published_at=entry.get("published", datetime.now().isoformat()),
                        category=feed_config["category"],
                    )
                )

            logger.debug(f"{feed_config['source']}에서 {len(entries)}건 수집")
        except Exception as e:
            logger.warning(f"RSS 수집 실패: {feed_config['source']} - {e}")

    return articles


async def collect_all_news() -> list[NewsArticle]:
    """전체 뉴스를 수집하고 중복을 제거

    네이버 경제/정치 뉴스와 글로벌 RSS 피드를 병렬로 수집합니다.

    Returns:
        중복 제거된 전체 뉴스 기사 목록
    """
    logger.info("뉴스 수집 시작...")

    economy_news, politics_news, rss_news = await asyncio.gather(
        _fetch_naver_section_news("101", "경제", "economy"),
        _fetch_naver_section_news("100", "정치", "politics"),
        _fetch_rss_news(),
    )

    all_news = [*economy_news, *politics_news, *rss_news]

    # 중복 제거 (제목 앞 30자 기준)
    seen: set[str] = set()
    unique: list[NewsArticle] = []
    for article in all_news:
        key = article.title[:30]
        if key not in seen:
            seen.add(key)
            unique.append(article)

    logger.info(f"총 {len(unique)}건 뉴스 수집 완료")
    return unique

File: src/collectors/news-collector.ts
import axios from 'axios';
import * as cheerio from 'cheerio';
import RssParser from 'rss-parser';
import { NewsArticle, NewsCategory } from '../types';
import logger from '../utils/logger';

const rssParser = new RssParser();

// ============================================
// 뉴스 수집기
// ============================================

/** 네이버 경제 뉴스 헤드라인 크롤링 */
async function fetchNaverEconomyNews(): Promise<NewsArticle[]> {
  try {
    const url = 'https://news.naver.com/section/101'; // 경제 섹션
    const { data } = await axios.get(url, {
      headers: {
        'User-Agent':
          'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
      },
      timeout: 10000,
    });

    const $ = cheerio.load(data);
    const articles: NewsArticle[] = [];

    // 헤드라인 뉴스 추출
    $('div.sa_text').each((_, el) => {
      const titleEl = $(el).find('a.sa_text_title');
      const title = titleEl.find('strong').text().trim();
      const href = titleEl.attr('href') || '';
      const summary = $(el).find('div.sa_text_lede').text().trim();

      if (title) {
        articles.push({
          title,
          summary: summary.slice(0, 200),
          url: href,
          source: '네이버 경제',
          publishedAt: new Date().toISOString(),
          category: categorizeNews(title + ' ' + summary),
        });
      }
    });

    logger.info(`네이버 경제 뉴스 ${articles.length}건 수집`);
    return articles.slice(0, 15);
  } catch (error) {
    logger.error('네이버 뉴스 수집 실패', { error });
    return [];
  }
}

/** 네이버 정치 뉴스 헤드라인 크롤링 */
async function fetchNaverPoliticsNews(): Promise<NewsArticle[]> {
  try {
    const url = 'https://news.naver.com/section/100'; // 정치 섹션
    const { data } = await axios.get(url, {
      headers: {
        'User-Agent':
          'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36',
      },
      timeout: 10000,
    });

    const $ = cheerio.load(data);
    const articles: NewsArticle[] = [];

    $('div.sa_text').each((_, el) => {
      const titleEl = $(el).find('a.sa_text_title');
      const title = titleEl.find('strong').text().trim();
      const href = titleEl.attr('href') || '';
      const summary = $(el).find('div.sa_text_lede').text().trim();

      if (title) {
        articles.push({
          title,
          summary: summary.slice(0, 200),
          url: href,
          source: '네이버 정치',
          publishedAt: new Date().toISOString(),
          category: 'politics',
        });
      }
    });

    logger.info(`네이버 정치 뉴스 ${articles.length}건 수집`);
    return articles.slice(0, 10);
  } catch (error) {
    logger.error('네이버 정치 뉴스 수집 실패', { error });
    return [];
  }
}

/** RSS 피드에서 글로벌 경제 뉴스 수집 */
async function fetchRssNews(): Promise<NewsArticle[]> {
  const feeds = [
    {
      url: 'https://feeds.bbci.co.uk/news/business/rss.xml',
      source: 'BBC Business',
      category: 'global' as NewsCategory,
    },
    {
      url: 'https://rss.nytimes.com/services/xml/rss/nyt/Economy.xml',
      source: 'NYT Economy',
      category: 'global' as NewsCategory,
    },
    {
      url: 'https://search.cnbc.com/rs/search/combinedcms/view.xml?partnerId=wrss01&id=10001147',
      source: 'CNBC',
      category: 'global' as NewsCategory,
    },
  ];

  const articles: NewsArticle[] = [];

  for (const feed of feeds) {
    try {
      const parsed = await rssParser.parseURL(feed.url);
      const items = (parsed.items || []).slice(0, 5);

      for (const item of items) {
        articles.push({
          title: item.title || '',
          summary: (item.contentSnippet || item.content || '').slice(0, 200),
          url: item.link || '',
          source: feed.source,
          publishedAt: item.isoDate || new Date().toISOString(),
          category: feed.category,
        });
      }

      logger.debug(`${feed.source}에서 ${items.length}건 수집`);
    } catch (error) {
      logger.warn(`RSS 수집 실패: ${feed.source}`, { error });
    }
  }

  return articles;
}

/** 뉴스 제목+내용으로 카테고리 자동 분류 */
function categorizeNews(text: string): NewsCategory {
  const lower = text.toLowerCase();

  if (/금리|기준금리|한은|연준|fed|통화정책|인플레/.test(lower)) return 'policy';
  if (/국회|대통령|정부|총리|외교|북한|선거/.test(lower)) return 'politics';
  if (/반도체|ai|배터리|자동차|it|테크|기술/.test(lower)) return 'industry';
  if (/코스피|코스닥|나스닥|s&p|dow|증시|주가/.test(lower)) return 'market';
  if (/미국|중국|일본|유럽|글로벌|세계/.test(lower)) return 'global';
  if (/트렌드|mz|소비|유행|핫/.test(lower)) return 'trend';
  return 'economy';
}

/** 전체 뉴스 수집 (메인 함수) */
export async function collectAllNews(): Promise<NewsArticle[]> {
  logger.info('뉴스 수집 시작...');

  const [economyNews, politicsNews, rssNews] = await Promise.all([
    fetchNaverEconomyNews(),
    fetchNaverPoliticsNews(),
    fetchRssNews(),
  ]);

  const allNews = [...economyNews, ...politicsNews, ...rssNews];

  // 중복 제거 (제목 기준)
  const seen = new Set<string>();
  const unique = allNews.filter((article) => {
    const key = article.title.slice(0, 30);
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });

  logger.info(`총 ${unique.length}건 뉴스 수집 완료`);
  return unique;
}
